src/main/java/org/traccar/helper/IpRetriever.java

```diff
@@ -0,0 +1,50 @@
+package org.traccar.helper;
+
+import javax.servlet.http.HttpServletRequest;
+
+/**
+ * Gets the client's IP address regardless of whether
+ * the server is behind a proxy/reverse proxy server or a load balancer.
+ */
+public final class IpRetriever {
+
+    /**
+     * Retrieves the client's IP address.
+     * Handles the cases like whether the server is behind a proxy/reverse proxy server or a load balancer
+     *
+     * @param request {@link HttpServletRequest} instance
+     * @return client's IP address
+     */
+    public static String retrieveIP(HttpServletRequest request) {
+
+        if (request != null) {
+            String ipAddress = request.getHeader("X-FORWARDED-FOR");
+
+            if (ipAddress != null && !ipAddress.isEmpty()) {
+                return removeUnwantedData(ipAddress);
+            } else {
+                ipAddress = request.getRemoteAddr();
+                return ipAddress;
+            }
+
+        } else {
+            return null;
+        }
+
+    }
+
+    /**
+     * If the server is behind a reverse proxy, the header value will also
+     * contain the IP's from load balancer and reverse proxy.
+     * This method gets rid of them.
+     *
+     * @param ipAddress IP address value from the header
+     * @return IP address of the client
+     */
+    private static String removeUnwantedData(String ipAddress) {
+        return ipAddress.contains(",") ? ipAddress.split(",")[0] : ipAddress;
```

Comment:
If you just need first value before comma, this can be much more efficient without using split.

Reply:
Alright, but in my opinion having such code without any comments as well would lead to confusion since it is not obvious why we would even need such an operation in the first place.

Reply:
What exactly can be more efficient?

Reply:
`indexOf(",")`